0011-CIR-trust-distribution.md

```diff
@@ -0,0 +1,116 @@
+# Dapr Trust Distribution
+
+* Author(s): joshvanl
+* State: Draft
+* Updated: 2023-10-24
+
+## Overview
+
+This is a design proposal to implement a proper trust distribution process in Dapr.
+Trust distribution will be implemented in a seamless way without downtime.
+This will improve and unlock security related features.
+
+## Background
+
+### Motivation
+
+To support the creation of features:
+
+- Proper Certificate Authority (CA) rotation (without re-using the root's private key)
+- External CA sources such as cert-manager and cloud provider CAs etc.
+- Dapr multi-cluster and networking federation
+
+Related issues:
+- [Multicluster Kubernetes](https://github.com/dapr/dapr/issues/3460)
+- [[Proposal] Support third-party CA - Integrate Cert Manager with Dapr](https://github.com/dapr/dapr/issues/3968)
+- [[Proposal] Automatic root certificate rotation](https://github.com/dapr/dapr/issues/5958)
+
+### Goals
+
+- Implement an active trust distribution mechanism for Dapr in Kubernetes that responds to updates.
+- Enable root rotation with no downtime.
+
+### Non-Goals
+
+- Sentry implements CA root rotation.
+- Implement external CA support.
+- Implement Dapr trust federation.
+
+### Current Shortfalls
+
+Trust distribution is the act of propagating trust data to enable secure communication or networking between peers.
+In the case of Dapr, this involves propagating PEM encoded CA bundle files to clients and servers in the cluster, which are then used to authenticate peers over TLS.
+Today, their are two methods of CA deployment to Dapr; either Sentry generated or provided by the Dapr cluster administrator.
+From the prospective of trust distribution, these two modes are functionally the same as they both result in the `dapr-trust-bundle` ConfigMap containing the CA bundle, and the `dapr-trust-bundle` Secret containing the issuer certificate chain.
+From here, trust distribution for the control plane occurs by the Operator, Placement, and Injector services reading from the mounted `dapr-trust-bundle` ConfigMap.
+Trust distribution for Daprds occurs from the Injector patching the Daprd container with an environment variable containing the CA Bundle originating from the `dapr-trust-bundle` ConfigMap.
+
+The problem with the current strategy is that once trust is distributed once (the `dapr-trust-bundle` ConfigMap and Secret is populated), the root of trust cannot change in the cluster.
```

Comment:
There's also the problem of rotating the root CA, which is one of the goals above?